### PR TITLE
Add service scheduler for adapter applications

### PR DESCRIPTION
**Pull request by alexwlchan**

### CODE REVIEW

File: lambdas/publish_to_sns.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    """
+    sns = boto3.client('sns')
+    resp = sns.publish(
+        TopicArn=topic_arn,
+        MessageStructure='json',
+        Message=json.dumps({
+            'default': json.dumps(**kwargs)
+        })
+    )
+    print(resp)
```
Comment (jamesgorrie) [inline lambdas/publish_to_sns.py]: Not worth logging here?

Comment (alexwlchan) [inline lambdas/publish_to_sns.py]: I think stdout is logged to CloudWatch, unless you had something else in mind?

Comment (jamesgorrie) [inline lambdas/publish_to_sns.py]: Nope, didn't realise that.

**Comment by alexwlchan** [comment]
Still to do:

- [ ] How to bundle the ZIP files?
- [ ] CloudWatch trigger for the first Lambda

File: lambdas/publish_to_sns.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    """
+    Given a topic ARN and a series of key-value pairs, publish the key-value
+    data to the SNS topic.
+    """
+    sns = boto3.client('sns')
+    resp = sns.publish(
+        TopicArn=topic_arn,
+        MessageStructure='json',
+        Message=json.dumps({
+            'default': json.dumps(kwargs)
```
Comment (kenoir) [inline lambdas/publish_to_sns.py]: Should we be more specific about the nature of the action sent here? This is essentially a scale up request so maybe we should standardise the request format and send that here?

Something like:

```json
{ 
  desiredCount: 3
  serviceName: "my_service"
}
```

**Comment by alexwlchan** [comment]
Okay, I think this works and is ready for review. A brief summary:

We have services that we only want to run for short periods. We’re managing this by changing the desired size of the associated ECS service tasks. It looks like this:

scheduler

When this runs automatically:

* We have scheduled CloudWatch events that trigger an AWS Lambda *publish_to_sns*
* This Lambda puts a "service schedule", aka a blob of JSON, on to an SNS topic
* The SNS topic triggers a second Lambda *update_ecs_service_size*
* That Lambda uses the ECS API to set the size of the ECS task

When the task is done, it writes to the same SNS topic with a desired size of 0, and ECS will proceed to tear it down. cf https://github.com/wellcometrust/platform-api/pull/84

---

The one problem I haven’t solved properly yet is getting the Python source into a Lambda.

You can provide source code to a Lambda in two ways: the inline code editor, or as a ZIP file. The latter can be a direct upload or come from an S3 bucket. You use the latter if you have external dependencies or a particularly large function; we have neither. I’m manually building a ZIP file that contains the file (OS X Finder > Compress *nameoffile.py*), but it would be nice to cut out the manual step – either have Terraform pass over the source directly, or construct the ZIP itself.

**Comment by kenoir** [comment]
You could generate the zip files with https://www.terraform.io/docs/providers/archive/d/archive_file.html

**Comment by alexwlchan** [comment]
Okay, I’ve switched to using `archive_file`. I haven’t tested it with a `terraform apply`, but I can see the ZIP files being created and that they have the correct contents.

File: terraform/lambdas.tf [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+  Lambda for updating ECS service size.
+
+  This is triggered by updates to an SNS topic.
+ */
+
+data "archive_file" "update_ecs_service_size" {
+  type        = "zip"
+  source_file = "../lambdas/update_ecs_service_size.py"
+  output_path = "/tmp/update_ecs_service_size.zip"
+}
```
Comment (alexwlchan) [inline terraform/lambdas.tf]: There’s an argument for moving this stanza inside the module – depends on whether we think we’ll be writing Lambdas that aren’t single-file Python scripts any time soon.

**Comment by alexwlchan** [comment]
Rebased and tidied.

File: terraform/lambdas.tf [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+EOF
+}
+
+resource "aws_sns_topic_subscription" "topic_lambda" {
+  topic_arn = "${aws_sns_topic.service_scheduler_topic.arn}"
+  protocol  = "lambda"
+  endpoint  = "${module.update_ecs_service_size_lambda.arn}"
+}
+
+resource "aws_lambda_permission" "with_sns" {
```
Comment (alexwlchan) [inline terraform/lambdas.tf]: The permissions are Lambda-specific – one allows SNS to invoke the Lambda, the other allows CloudWatch to invoke the Lambda.

File: terraform/lambdas.tf [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+  input = <<EOF
+{
+  "topic_arn": "${aws_sns_topic.service_scheduler_topic.arn}",
+  "cluster": "${aws_ecs_cluster.services.name}",
+  "service": "calm_adapter",
+  "desired_count": 1
+}
+EOF
+}
```
Comment (alexwlchan) [inline terraform/lambdas.tf]: I think in this case, it makes sense to have each Lambda in a separate file. They won’t always have the same resources.

File: terraform/lambda/role.tf [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+      "Effect": "Allow",
+      "Sid": ""
+    }
+  ]
+}
+EOF
+}
+
+resource "aws_iam_role_policy" "lambda_cloudwatch_logs" {
+  name = "lambda_cloudwatch_logs"
```
Comment (alexwlchan) [inline terraform/lambda/role.tf]: Interpolate the name of the Lambda

**Comment by alexwlchan** [comment]
@kenoir I think this is ready for review. In particular commits 4–6, which modularise the triggers and break up the Lambdas into different files.

File: terraform/lambda/sns.tf [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+}
+
+resource "aws_sns_topic_subscription" "topic_lambda" {
+  count     = "${var.sns_trigger_arn == "" ? 1 : 0}"
+  topic_arn = "${var.sns_trigger_arn}"
+  protocol  = "lambda"
+  endpoint  = "${aws_lambda_function.lambda_function.arn}"
+}
+
+*/
```
Comment (kenoir) [inline terraform/lambda/sns.tf]: Should this all be commented out?

File: terraform/lambda/trigger_cloudwatch/cloudwatch.tf [hunk 8]
```diff
@@ -0,0 +1,13 @@
+resource "aws_lambda_permission" "allow_cloudwatch_to_call_check_foo" {
```
Comment (kenoir) [inline terraform/lambda/trigger_cloudwatch/cloudwatch.tf]: Bar?

File: terraform/lambda/variables.tf [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+  description = "Description of the Lambda function"
+}
+
+variable "filename" {
+  description = "Path to the Python file containing the Lambda source code"
+}
+
+variable "sns_trigger_arn" {
+  description = "If this Lambda is triggered by SNS, the ARN of the trigger topic"
+  default     = ""
```
Comment (kenoir) [inline terraform/lambda/variables.tf]: Should these have no defaults, rather than empty string?

**Comment by alexwlchan** [comment]
Okay, think the markups are done.

**Pull request merged**